core-plugins/src/main/java/io/cdap/plugin/batch/source/FTPBatchSource.java

```diff
@@ -181,15 +212,63 @@ public Schema getSchema() {
       return SCHEMA;
     }
 
+    public String extractPasswordFromUrl() {
+      int getLastIndexOfAtSign = path.lastIndexOf("@");
+      String authentication = path.substring(0, getLastIndexOfAtSign);
+      return authentication.substring(authentication.lastIndexOf(":") + 1);
+    }
+
+    public boolean authContainsSpecialCharacters() {
+      Pattern regularPasswordWithoutSpecialCharacters = Pattern.compile("[^A-Za-z0-9]");
+      Matcher regularPassword = regularPasswordWithoutSpecialCharacters.matcher(extractPasswordFromUrl());
+      return !regularPassword.matches();
+    }
+
     Map<String, String> getFileSystemProperties() {
-      if (fileSystemProperties == null) {
+      if (fileSystemProperties == null && !authContainsSpecialCharacters()) {
```

Comment:
I think we need this condition because when authentication in url contains special characters but File System Properties are empty, we'll return always an empty hashmap at the end of the function but this doesn't solve our problem. 

Instead, we need to extract the necessary parameters from **url** (host, port, username, password) when this scenario happens and add these parameters as File System Properties in order to solve the issue.

Reply:
If  authentication in url contains special characters but File System Properties are empty, we will go into the if condition in line 237 and update fileSystemPropertiesMap.

Reply:
You're right. Now I got your point. We don't need that if condition. I followed your suggestion and tested the scenarios after removing the condition and everything is working okay.